File: cpp/exercises/trie.cpp
#include "trie.h"
 
// ─── TrieNode ──────────────────────────────────────────────────────────────
//
// TODO: implement TrieNode.
//
// A TrieNode needs:
//   - children: std::unordered_map<char, TrieNode*>
//               Maps each character to its child node. Initialize as empty map.
//   - isEndOfWord: bool
//               True if an inserted word ends at this node. Default: false.
//
// Write a constructor that initializes isEndOfWord = false.
// (children will default-initialize to an empty map automatically.)
//
struct Trie::TrieNode {
    std::unordered_map<char, TrieNode*> children;
    bool isEndOfWord;
    // TODO: implement constructor
      TrieNode() : isEndOfWord(false) {}
};
 
 
// ─── Destructor (pre-implemented) ─────────────────────────────────────────
 
void Trie::destroyNode(TrieNode* node) {
    if (!node) return;
    for (auto& [ch, child] : node->children) {
        destroyNode(child);
    }
    delete node;
}
 
Trie::~Trie() {
    destroyNode(root_);
}
 
 
// ─── Constructor ──────────────────────────────────────────────────────────
 
Trie::Trie() {
    // TODO: implement this
    // Initialize root_ to a freshly allocated TrieNode.
    // root_ = new TrieNode();
     root_ = new TrieNode();// placeholder — replace with the line above
    //
    // Note: body assignment (not an initializer list) is used here because
    // TrieNode is an incomplete type at the point of Trie's declaration.
    // It's only fully defined below in this .cpp file.
}
 
 
// ─── insert ───────────────────────────────────────────────────────────────
 
void Trie::insert(const std::string& word) {
    // TODO: implement this
    // Walk from root_, creating TrieNodes as needed for each character.
    // After all characters, set the final node's isEndOfWord = true.
    TrieNode* current = root_;
 
    for (char ch : word) {
        if (current->children.find(ch) == current->children.end()) {
            current->children[ch] = new TrieNode();
        }
        current = current->children[ch];
    }
 
    current->isEndOfWord = true;
}
 
 
// ─── search ───────────────────────────────────────────────────────────────
 
bool Trie::search(const std::string& word) const {
    // TODO: implement this
    // Walk from root_ following each character. If a child doesn't exist,
    // return false. At the end, return the node's isEndOfWord.
    if(!root_) {
        return false;
    }    

    TrieNode* current = root_;
 
    for (char ch : word) {
        if (current->children.find(ch) == current->children.end()) {
            return false;
        }
        current = current->children.at(ch);
    }
 
    return current->isEndOfWord;// placeholder
}
 
 
// ─── startsWith ───────────────────────────────────────────────────────────
 
bool Trie::startsWith(const std::string& prefix) const {
    // TODO: implement this
    // Walk from root_ following each character of prefix. If a child doesn't
    // exist, return false. If you reach the end of prefix, return true.
    if(!root_) {
        return false;
    }    


     TrieNode* current = root_;
 
    for (char ch : prefix) {
        if (current->children.find(ch) == current->children.end()) {
            return false;
        }
        current = current->children.at(ch);
    }
 
    return true;
}
 

File: java/src/main/java/exercises/Trie.java
package exercises;

import java.util.HashMap;
import java.util.Map;

/**
 * Exercise: Trie (Prefix Tree)
 * ================================
 * A trie stores strings by sharing common prefixes. Each node represents
 * one character. Walking from root to a node spells out a prefix. The
 * isEndOfWord flag marks where full words end.
 *
 * Example: inserting "app" and "apple" shares nodes for 'a', 'p', 'p'.
 * The second 'p' node has isEndOfWord = true (end of "app"), and the 'e'
 * node eventually has isEndOfWord = true (end of "apple").
 *
 * All operations are O(m) where m is the length of the word/prefix.
 *
 * Your task: implement TrieNode first, then the constructor and all methods
 * marked with TODO.
 */
public class Trie {

    /**
     * A single node in the trie.
     *
     * You need to implement this class. A TrieNode needs:
     *   - children: Map&lt;Character, TrieNode&gt; — maps each character to its
     *               child node. Initialize as a new HashMap.
     *   - isEndOfWord: boolean — true if an inserted word ends at this node.
     *                  Default: false
     *
     * Hint: write a no-arg constructor that initializes both fields.
     */
    public static class TrieNode {
        // TODO: declare fields (Map<Character, TrieNode> children, boolean isEndOfWord)
        // TODO: implement constructor
    }

    private TrieNode root;

    /**
     * Initialize the trie with an empty root TrieNode.
     * Start here — this is the first method to implement.
     */
    public Trie() {
        // TODO: initialize root
        throw new UnsupportedOperationException("Not implemented yet");
    }

    /**
     * Insert word into the trie.
     *
     * Walk from root, creating TrieNodes as needed for each character.
     * After processing all characters, mark the final node's isEndOfWord = true.
     */
    public void insert(String word) {
        // TODO: implement this
        throw new UnsupportedOperationException("Not implemented yet");
    }

    /**
     * Return true if word was previously inserted, false otherwise.
     *
     * Walk from root following each character. If a character's child doesn't
     * exist, return false. At the end, return isEndOfWord of the final node.
     *
     * Think about: why is isEndOfWord necessary? What goes wrong if you just
     * check whether the last node exists?
     */
    public boolean search(String word) {
        // TODO: implement this
        throw new UnsupportedOperationException("Not implemented yet");
    }

    /**
     * Return true if any inserted word starts with prefix.
     *
     * Walk from root following each character of prefix. If a character's
     * child doesn't exist, return false. If you reach the end of the prefix,
     * return true — isEndOfWord doesn't matter here.
     */
    public boolean startsWith(String prefix) {
        // TODO: implement this
        throw new UnsupportedOperationException("Not implemented yet");
    }
}
